kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,118 @@ struct LogReplayScanner {
     seen: HashSet<FileActionKey>,
 }
 
-#[derive(Default)]
-struct AddRemoveVisitor {
-    adds: Vec<(Add, usize)>,
-    removes: Vec<Remove>,
-    selection_vector: Option<Vec<bool>>,
-    // whether or not we are visiting commit json (=true) or checkpoint (=false)
+/// A visitor that deduplicates a stream of add and remove actions into a stream of valid adds. Log
+/// replay visits actions newest-first, so once we've seen a file action for a given (path, dvId)
+/// pair, we should ignore all subsequent (older) actions for that same (path, dvId) pair. If the
+/// first action for a given file is a remove, then that file does not show up in the result at all.
+struct AddRemoveDedupVisitor<'seen> {
+    seen: &'seen mut HashSet<FileActionKey>,
+    selection_vector: Vec<bool>,
     is_log_batch: bool,
 }
 
-const ADD_FIELD_COUNT: usize = 15;
+impl<'seen> AddRemoveDedupVisitor<'seen> {
+    /// Checks if log replay already processed this logical file (in which case the current action
+    /// should be ignored). If not already seen, register it so we can recognize future duplicates.
+    fn already_seen(&mut self, key: FileActionKey) -> bool {
```

Comment:
can we rename this as `record_seen`? I prefer to note the side-effect in the name since it makes things like the order of things like `!self.already_seen(..) && is_add` important. `already_seen` sounds like a check, so you'd think it should be okay to reorder that and avoid a probe, but obviously it's not

Reply:
Went with `check_and_record_seen` to be clear that it's doing both